### PR TITLE
Unreachable Select Case Inspection

### PR DESCRIPTION
**Pull request by BZngr**
Related to #1584 and #3050 

1. Detects unreachable Cases Clauses caused by duplicate, invalid comparison-types,  or prior super-set conditions.
2. Detects unreachable Case Else clauses where all possible cases are previously handled.
3. Clauses that contain invalid comparison types are not flagged if there are other valid comparison types present within the same Case Clause.
4. Attempts to makes comparisons of mixed types that would be implicitly converted and evaluated by VBA.
5. Supports evaluation of enums.  Enums Select Case statements will not mark the Case Else clause as unreachable if all enum values are handled.  Enums will be inspected in the same manner as Longs.  Case Else will be flagged as unreachable if and only if all Long values are handled by prior Case Clauses.
6. Will pass on inspecting Select Case statements where the Select Case statement includes math on the Select Case variable (e.g.,` Select Case x ^ 2 / 4`).

### CODE REVIEW

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 1]
```diff
@@ -0,0 +1,1309 @@
...
+
+        public override IInspectionListener Listener { get; } =
+            new UnreachableCaseInspectionListener();
+
+        public override CodeInspectionType InspectionType => CodeInspectionType.CodeQualityIssues;
+
+        private static Dictionary<string, string> CompareInversions = new Dictionary<string, string>()
+        {
+            { CompareTokens.EQ, CompareTokens.EQ },
+            { CompareTokens.NEQ, CompareTokens.NEQ },
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: why is NEQ the inverse of NEQ and EQ the inverse of EQ? Shouldn't it be:
```
{ CompareTokens.EQ, CompareTokens.NEQ },
{ CompareTokens.NEQ, CompareTokens.EQ },
```

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 2]
```diff
@@ -0,0 +1,1309 @@
...
+            { CompareTokens.LT, CompareTokens.GT },
+            { CompareTokens.LTE, CompareTokens.GTE },
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Truly inverse comparisons would be `{ LT, GTE }, { LTE, GT }`

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 3]
```diff
@@ -0,0 +1,1309 @@
...
+            }
+
+            var expression = ParserRuleContextHelper.GetDescendent<SimpleNameExprContext>(firstLExpr).GetText();
+            if (SymbolList.TypeHintToTypeName.ContainsKey(expression.Last().ToString()))
+            {
+                return SetTheTypeNames(selectStmtDO, SymbolList.TypeHintToTypeName[expression.Last().ToString()]);
+            }
+
+            var idRefs = (State.DeclarationFinder.MatchName(expression).Select(dec => dec.References))
+                .SelectMany(rf => rf).Where(idr => ParserRuleContextHelper.HasParent(idr.Context, selectStmtDO.SelectExpressionContext));
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: ParserRuleContextHelper seems like it should be extension methods... `idr.Context.HasParent(selectStmtDO.SelectExpressionContext)` would be a little cleaner

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 4]
```diff
@@ -0,0 +1,1309 @@
...
+                if (caseClause.ResultType == ClauseEvaluationResult.NoResult)
+                {
+                    caseClause = InspectCaseClause(caseClause, selectStmtDO.SummaryCaseClauses);
+                    selectStmtDO = UpdateSummaryClauses(selectStmtDO, caseClause);
+
+                    if (caseClause.RangeClauseDOs.All(rg => rg.ResultType != ClauseEvaluationResult.NoResult))
+                    {
+                        caseClause.ResultType = caseClause.RangeClauseDOs.All(rg => rg.ResultType == ClauseEvaluationResult.Unreachable)
+                            ? ClauseEvaluationResult.Unreachable : caseClause.RangeClauseDOs.All(rg => rg.ResultType == ClauseEvaluationResult.MismatchType)
+                                ? ClauseEvaluationResult.MismatchType : ClauseEvaluationResult.NoResult;
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Could this be more explicit and less ternaries, please? I find it hard to follow this

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 5]
```diff
@@ -0,0 +1,1309 @@
...
+            }
+            return caseClauseDO;
+        }
+
+        private RangeClauseDataObject InitializeRangeClauseDataObject(RangeClauseDataObject rangeClauseDO, string targetTypeName, string refName )
+        {
+            rangeClauseDO.TypeNameTarget = targetTypeName;
+            rangeClauseDO.IdReferenceName = refName;
+            rangeClauseDO.UsesIsClause = HasChildToken(rangeClauseDO.Context, Tokens.Is);
+            rangeClauseDO.IsValueRange = HasChildToken(rangeClauseDO.Context, Tokens.To);
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: HasChildToken should also be an extension method

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 6]
```diff
@@ -0,0 +1,1309 @@
...
+                    if (!parentData.EvaluateAsIsClause)
+                    {
+                        parentData.EvaluateAsIsClause = CompareOperations.Keys.Contains(child.GetText());
+                    }
+                    parentData.Operator = child.GetText();
+                    contextEvals = AddEvaluationData(contextEvals, parentContext, parentData);
+                    continue;
+                }
+
+                if (IsBinaryOperatorContext(child) || IsUnaryOperandContext( child) )
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: extra space after opening paren

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 7]
```diff
@@ -0,0 +1,1309 @@
...
+                        SelectCaseRefName = valueDeclaration.IdentifierName
+                    };
+
+                    parentData = AddEvaluationData(parentData, child, exprEval);
+                    return ResolveContextValue(parentData, child).First().Value.Result.AsString();
+                }
+
+                if (child is LiteralExprContext)
+                {
+                    if(child.Parent is EnumerationStmt_ConstantContext)
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: missing space before opening paren

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 8]
```diff
@@ -0,0 +1,1309 @@
...
+                }
+            }
+            return string.Empty;
+        }
+
+        private List<ParserRuleContext> GetRHSContexts(List<IParseTree> contexts)
+        {
+            var contextsOfInterest = new List<ParserRuleContext>();
+            var eqIndex = contexts.FindIndex(ch => ch.GetText().Equals(CompareTokens.EQ));
+            if(eqIndex == contexts.Count)
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: missing space before paren

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 9]
```diff
@@ -0,0 +1,1309 @@
...
+            else
+            {
+                summaryClauses.SingleValues.Add(rangeClauseDO.SingleValue);
+            }
+            return summaryClauses;
+        }
+
+        private bool EvaluateCaseElseAccessibility(SummaryCaseCoverage summaryClauses, string typeName)
+        {
+            if(summaryClauses.CaseElseIsUnreachable) { return summaryClauses.CaseElseIsUnreachable; }
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: missing opening paren

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs [hunk 10]
```diff
@@ -0,0 +1,1309 @@
...
+            if (summaryClauses.IsLTMax != null && summaryClauses.IsGTMin != null)
+            {
+                if (summaryClauses.IsLTMax > summaryClauses.IsGTMin
+                        || (summaryClauses.IsLTMax >= summaryClauses.IsGTMin
+                                && summaryClauses.SingleValues.Contains(summaryClauses.IsLTMax)))
+                {
+                    return true;
+                }
+
+                else if(summaryClauses.Ranges.Count > 0)
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: missing space before opening paren

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+            return thisValue._operatorIsLT(thisValue, compValue);
+        }
+
+        public static bool operator ==(UnreachableCaseInspectionValue thisValue, UnreachableCaseInspectionValue compValue)
+        {
+            if(ReferenceEquals(null, thisValue))
+            {
+                return ReferenceEquals(null, compValue);
+            }
+            else
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs]: opening the else block here is not necessary :)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+            }
+        }
+
+        public static bool operator !=(UnreachableCaseInspectionValue thisValue, UnreachableCaseInspectionValue compValue)
+        {
+            if (ReferenceEquals(null, thisValue))
+            {
+                return !ReferenceEquals(null, compValue);
+            }
+            else
```
Comment (Vogel612) [inline Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs]: opening the else block here is not necessary

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 1 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Yeah...I think a better name for the dictionary might be "AlgebraicLogicInversions".  The goal is to be able to perform some (very) limited algebra for logical Range Clause expressions like '6 < x'.  In order to move 'x' to the LHS, the desired results is x > 6.  So, for '6 = x' the algebra results desired is 'x = 6'.  Same for NEQ.  I'm definitely open to a better name.  Note: I am interested in moving the variable to the LHS so that the expression can be evaluated in the same way as an IS clause. ('x < 6' == 'Is < 6' when 'x' is the Select Case expression)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 2 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: See comment above - the goal is algebraic inversion rather than logical inversion.

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 3 above)
Comment (Hosch250) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Pretty sure that is an extension method already. If it isn't, Vogel's right.

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 2 above)
Comment (Hosch250) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Using C# 6, I'm pretty sure you could format these as `[CompareTokens.EQ] = CompareTokens.EQ`, etc.

Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: @Hosch250 yes!  Updated this and other Dictionaries accordingly

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 3 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Added extension method(s)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 4 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Modified...yes...more readable.

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 5 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Added extension method(s).  If the intent is to refactor ParserRuleContextHelper completely into extension methods, then I would like to make that a separate Issue/PR.  I left ParserRuleContextHelper as-is.

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 6 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Cleaned-up (and others like it)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 7 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Cleaned-up (and others like it)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 8 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Cleaned-up (and others like it)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs (hunk 9 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspection.cs]: Cleaned-up (and others like it)

File: Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs (hunk 11 above)
Comment (BZngr) [reply Rubberduck.Inspections/Concrete/UnreachableCaseInspectionValue.cs]: Cleaned-up (and others like it)

**Comment by bclothier** [comment]
@BZngr in regard to this point:

2. Detects unreachable Case Else clauses where all possible cases are previously handled.

I'm just curious - even if one had a function that takes an `Enum` as the parameter and the `Select Case` within the function enumerated all the values for that `Enum`, it is still valid to do this:

`MyFunction(1234) 'Look, a random number!`

which is implicitly converted but still won't match any of the `Case` statement. 

I often add a guard statement similar to this:

```
   Case Else
     Err.Raise 5, "MyFunction", "Invalid value given for the enumeration."
```

Will the inspection flag that as unreachable?

**Comment by BZngr** [comment]
Updated to remove strict enum evaluations for unreachable Case and CaseElse clauses.  Enums are now evaluated in the same manner as Longs.

**Comment by BZngr** [comment]
I'm closing this pull request and plan resubmit again later.  Even though it finds many scenarios of unreachable code, I believe it could use more refactoring/rework to make it easier to understand and review.

And a question:
It seems that most (if not all) other inspections seem  to 'fit' into a single inspection class/file.  So, I tried to do the same with this inspection - and still ended up with two classes/files.  To reduce the file/class count, the solution has several supporting Structs and processes them.  But, I think there may be read/review-ability benefits to refactoring some of the Structs into classes.  I'm thinking that the additional files should have their own folder (a subfolder of Inspections\Concrete)  to organize the supporting classes.  Since I do not see this organization/approach for the other existing inspections, I wanted to check if this was an acceptable way to organize my code.  Any guidance here would be appreciated - thanks.

**Comment by bclothier** [comment]
RE: the question - FWIW if you look at Extract Method, there are several files, way more than other refactoring normally needs. 

I wouldn't be surprised if we have other inspections where fitting everything in a one file don't make sense. I think that as long we follow the same convention used by refactoring -- creating a folder instead of a file, it should be OK to do that for inspections, too.

**Comment by wqweto** [comment]
@BZngr FYI, following is a valid VB code that does not work as intended:
```
Select Case x
  Case x ^ 2 = 49
    'OK
  Case x = 7
    'Unreachable, but not detected - math/algebra on the Select Case variable yet to be supported
  Case Is < 50
    ...
End Select
```
The expressions in first two `Case` statements are boolean so it makes sence to only `Select Case True` or `False`.

Basicly `Case x < 50` is different than `Case Is < 50`  for a `Select Case x` -- first one is probably wrong.

Noticed you are using similar in `UnreachableCaseInspectionTests`

**Comment by bclothier** [comment]
@wqweto  I don't know if this is appropriate for this inspection. This is definitely something we need to inspect but this strikes me as more about inspecting for implicit type conversion than for unreachable code because VBA allow us to do implicit voodoo conversions where we can write expressions that results in a boolean which is then converted to whatever `x` is (presumably an integer of some size), which would cause unexpected results as you said.  Furthermore this problem would not be limited to only the `Select Case` so for that reason, I think it's best as its own inspection, apart from this PR. I suggest that you open an issue if there's not an inspection already for that type of implicit type conversions in a comparison.

**Comment by bclothier** [comment]
BTW, see #3050
